File: docs/samples/hello.ts
// Greetings in a few languages, with enough TypeScript in it to show a colour theme.
type Language = "en" | "el" | "fr";

interface Greeting {
  readonly language: Language;
  readonly template: string;
}

const DEFAULT_TARGET = "World";
const MAX_WIDTH = 42;

class Greeter {
  private readonly greetings: readonly Greeting[];
  private readonly width: number;
  private count = 0;

  constructor(greetings: readonly Greeting[], width: number = MAX_WIDTH) {
    this.greetings = greetings;
    this.width = width;
  }

  get languages(): Set<Language> {
    return new Set(this.greetings.map((greeting) => greeting.language));
  }

  greet(target: string = DEFAULT_TARGET): string[] {
    const lines = this.greetings.map(({ template }) => template.replace("{target}", target));
    this.count += lines.length;
    if (lines.some((line) => line.length > this.width)) {
      throw new RangeError(`a greeting exceeds ${this.width} characters`);
    }
    return lines;
  }

  async greetLater(target?: string, delayMs = 10): Promise<string[]> {
    await new Promise<void>((resolve) => setTimeout(resolve, delayMs));
    return this.greet(target);
  }

  get total(): number {
    return this.count;
  }
}

async function main(argv: string[]): Promise<number> {
  const target = argv[2] ?? DEFAULT_TARGET;
  const greeter = new Greeter([
    { language: "en", template: "Hello, {target}!" },
    { language: "el", template: "Γεια σου, {target}!" },
    { language: "fr", template: "Bonjour, {target} !" },
  ]);
  const lines = await greeter.greetLater(target);
  lines.forEach((line, index) => console.log(`${String(index + 1).padStart(2)}. ${line}`));
  console.log(`-- ${greeter.total} greetings in ${greeter.languages.size} languages`);
  return 0;
}

main(process.argv).then((code) => process.exit(code));

File: docs/samples/hello.rs
//! Greetings in a few languages, with enough Rust in it to show a colour theme.
use std::fmt;

const DEFAULT_TARGET: &str = "World";
const MAX_WIDTH: usize = 42;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
enum Language {
    English,
    Greek,
    French,
}

#[derive(Debug, Clone)]
struct Greeting {
    language: Language,
    template: &'static str,
}

impl Greeting {
    fn render(&self, target: &str) -> String {
        self.template.replace("{target}", target)
    }
}

impl fmt::Display for Greeting {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}: {}", self.language, self.template)
    }
}

struct Greeter {
    greetings: Vec<Greeting>,
    width: usize,
    count: usize,
}

impl Greeter {
    fn new(greetings: Vec<Greeting>) -> Self {
        Self { greetings, width: MAX_WIDTH, count: 0 }
    }

    fn greet(&mut self, target: &str) -> Result<Vec<String>, String> {
        let lines: Vec<String> = self.greetings.iter().map(|g| g.render(target)).collect();
        if let Some(long) = lines.iter().find(|line| line.chars().count() > self.width) {
            return Err(format!("greeting exceeds {} characters: {long}", self.width));
        }
        self.count += lines.len();
        Ok(lines)
    }
}

fn main() -> Result<(), String> {
    let target = std::env::args().nth(1).unwrap_or_else(|| DEFAULT_TARGET.to_string());
    let mut greeter = Greeter::new(vec![
        Greeting { language: Language::English, template: "Hello, {target}!" },
        Greeting { language: Language::Greek, template: "Γεια σου, {target}!" },
        Greeting { language: Language::French, template: "Bonjour, {target} !" },
    ]);
    for (index, line) in greeter.greet(&target)?.iter().enumerate() {
        println!("{:>2}. {line}", index + 1);
    }
    println!("-- {} greetings", greeter.count);
    Ok(())
}
